language-plutus-core/src/Language/PlutusCore/Clone.hs

```diff
@@ -1,50 +1,163 @@
                                  ) where
 
-import           Control.Monad.State.Class
-import           Control.Monad.Trans.State hiding (get, modify)
-import qualified Data.IntMap               as IM
 import           Language.PlutusCore.Name
 import           Language.PlutusCore.Quote
 import           Language.PlutusCore.Type
-import           PlutusPrelude
-
-type CloneSt = IM.IntMap (TyNameWithKind ())
-
-cloneEnvAssign :: MonadState CloneSt m
-               => Unique
-               -> TyNameWithKind ()
-               -> m ()
-cloneEnvAssign (Unique i) ty = modify (IM.insert i ty)
-
-cloneType :: MonadQuote m => Type TyNameWithKind a -> m (Type TyNameWithKind ())
-cloneType = flip evalStateT mempty . cloneTypeM
-
-replaceTyName :: (MonadQuote m, MonadState CloneSt m) => TyNameWithKind () -> m (TyNameWithKind ())
-replaceTyName (TyNameWithKind (TyName (Name l' n u))) = do
-    u' <- liftQuote freshUnique
-    let tn = void $ TyNameWithKind (TyName (Name l' n u')) -- TODO use lenses
-    cloneEnvAssign u tn
-    pure tn
-
-cloneTypeM :: (MonadQuote m, MonadState CloneSt m) => Type TyNameWithKind a -> m (Type TyNameWithKind ())
-cloneTypeM (TyFix _ tn ty) = do
-    tn' <- replaceTyName (void tn)
-    TyFix () tn' <$> cloneTypeM ty
-cloneTypeM (TyForall _ tn k ty) = do
-    tn' <- replaceTyName (void tn)
-    TyForall () tn' (void k) <$> cloneTypeM ty
-cloneTypeM (TyLam _ tn k ty) = do
-    tn' <- replaceTyName (void tn)
-    TyLam () tn' (void k) <$> cloneTypeM ty
-cloneTypeM (TyApp _ ty ty') = TyApp () <$> cloneTypeM ty <*> cloneTypeM ty'
-cloneTypeM (TyFun _ ty ty') = TyFun () <$> cloneTypeM ty <*> cloneTypeM ty'
-cloneTypeM x@(TyVar _ (TyNameWithKind (TyName (Name _ _ u)))) = do
-    cloneSt <- get
-    case IM.lookup (unUnique u) cloneSt of
-        Just n  -> pure (TyVar () n)
-        Nothing -> pure (void x)
-cloneTypeM x@TyBuiltin{}    = pure (void x)
-cloneTypeM x@TyInt{}        = pure (void x)
+
+import           Control.Applicative
+import           Control.Monad.Reader
+import           Control.Monad.State
+
+import           Data.Foldable
+import           Data.Functor.Foldable
+import qualified Data.IntMap               as IM
+import qualified Data.IntSet               as IS
+import qualified Data.List.NonEmpty        as NE
+import           Data.Maybe                (fromMaybe)
+import           Lens.Micro
+
+type Cloning m = (MonadState CloneSt m, MonadReader CloneWhat m, MonadQuote m)
+
+data CloneSt = CloneSt {
+    csMappingStack :: NE.NonEmpty (IM.IntMap Unique) -- ^ A scoped stack of unique mappings.
+    , csBound      :: IS.IntSet -- ^ The set of bound variables seen so far.
+    }
+
+emptyState :: CloneSt
+emptyState = CloneSt {csMappingStack=mempty NE.:| [], csBound=mempty}
+
+addMapping :: Cloning m => Int -> Unique -> m ()
+addMapping oldU newU = modify $ \(CloneSt (top NE.:| rest) bound) -> CloneSt (IM.insert oldU newU top NE.:| rest) bound
```

Comment:
What does this stack do? The previous method with explicitly passing state seems better.

Reply:
I should write some more explanation - clearly it's not as obvious as I thought.

The idea here is to explicitly mirror the scopes. So we have one map per scope, and mappings are inserted in the scope where they are first seen. I think that makes it more obvious how we're handling scoping, whereas in the current renamer it's not at all clear to me how it handles scoping and shadowing. (the comment even describes it as "convoluted").

Reply:
Right, I'm primarily curious as to how this affects performance. Personally, I find it more difficult to follow, but my main concern is performance.